docs/api/version-history.md

```diff
@@ -50,6 +50,14 @@ keywords: "API, Docker, rcli, REST, documentation"
   daemon has experimental features enabled.
 * `GET /networks/{id}` now returns an `EnableIPv4` field showing whether the
   network has IPv4 IPAM enabled.
+* `POST /networks/{id}/connect` and `POST /containers/create` now accept a
+  `Priority` field in `EndpointsConfig`. This value is used to determine which
+  network endpoint provides the default gateway for the container. The endpoint
+  with the highest priority is selected. If multiple endpoints have the same
+  priority, endpoints are sorted lexicographically by their network name, and
+  the one that sorts first is picked.
```

Comment:
Silly question (ISTR @robmry also mentioned something along those lines); if the intent of this option is to determine the default gateway, would it make sense to instead consider adding a new option to define which network should be used for this?

Basically considering that the order itself was a workaround for not being able to configure the default; it seems like now we're improving the workaround, but perhaps we could implement the actual thing that it's used for somehow?

Reply:
Also curious what the behavior is when connecting to a new network with a higher priority; will it update the existing config to reconfigure the default gateway, or does this only happen when the container is initially created?

Reply:
> if the intent of this option is to determine the default gateway

That's what it is.

> would it make sense to instead consider adding a new option to define which network should be used for this?

Such a new option would have a major drawback: since you can't update a container, your container would have no gateway and loose access to the outside world when you'd disconnect it from the network providing the default gateway.

Another drawback: you couldn't use `docker network connect` to attach a network that should become the default gateway.

> Also curious what the behavior is when connecting to a new network with a higher priority; will it update the existing config to reconfigure the default gateway

Yup, that's what's happening -- libnetwork already does that automatically when the gateway changes for any reason (eg. because the container was not connected to a non-internal network, or because the network getting connected sorts before in lexicographic order).

This gives the ability to end-users to choose which network should be the default gateway without worrying about the order of calls made to the `NetworkConnect` endpoint.

On `NetworkDisconnect` the endpoint with the highest priority or sorting first in lexicographic order will be used.

> Basically considering that the order itself was a workaround for not being able to configure the default

You mean, the lexicographic sorting? It's not going away, but with this change it's neither the only nor the recommended way to set the default gateway.

Reply:
Yeah, so considering that (and this would already be the current situation); in the following case, `project_network` would become the default gateway; this would (e.g.) be the project's default network used in a compose-stack;

```bash
docker network connect project_network container
```

But if the container now is later connected to `global_network`, that one (sorting before `project_network`) would now become the default gateway; this can be a common scenario for users using (e.g.) traefik or nginx-proxy as proxy to map domains;

```bash
docker network connect global_network container
```

So to counteract that, I would need to disconnect from the `project_network`, find a value that's higher than any of the existing networks it's connected to, and re-connect to that network. While a `use this network as default gateway` option is not ideal, it may give a slightly stronger indication "this is what to use";

- if no connected network as `use this network as default gateway` option is set, we would still use the current (sort networks by name)
- if a network has the option set though, then we use that as the default (priority over any other ordering)
- ❓ we could disallow connecting if there's already a network connected that's configured as default gateway
- if that network would be disconnected, we'd probably have to fall back to the "sort by name" approach; I _could_ see use-cases where disconnecting would mean "no gateway", but that would be trickier to implement probably.

Reply:
A priority value seems better to me - because it's clear what will happen when the current gateway network is disconnected, or a new/desired-gateway network is connected to a running container.

If a specific network should always be the gateway, only that network will need a priority value. In that case, setting "priority=1" is equivalent to saying "this is the gateway" - no more or less confusing, but the corner cases or wanting to switch gateway networks are easier to deal with. (A fixed choice of gateway network probably the common case though - after not having a choice of gateway networks, or it not mattering which network is used.)

In the global/project_network example ... if the user wants to keep `project_network` as the gateway, they can either give it a high priority when creating the container, then not specify a priority for other networks. Or, specify a negative priority (I think, maybe worth a testcase?) when connecting `global_network`. If they want `global_network` to take over, they just use a higher priority.

Reply:
If I want a dual stack network acting as the IPv4 gateway, and an IPv6 only network acting as the v6 gateway ...

At the moment, even with the new priority field, the dual stack network will be the gateway [for both](https://github.com/moby/moby/blob/7b8b63ec2219d88edb21b6e06054384a01ad76d9/libnetwork/default_gateway.go#L195). But, with priority plumbed in we could/should change that, and pick the v4 and v6 endpoints separately.

Then, if I give the v6 network higher priority than the dual stack, and the dual stack network higher priority than any other networks - it'll do what I want.

We can make that change in a separate PR - but I think it strengthens the case for priority values. Without them, we'd need separate options for selecting v4 and v6 gateways.

Reply:
> Although ... if I have two dual-stack networks, and want one to be the IPv4 gateway and the other the IPv6 gateway - we can't describe that with a priority field. It'd have to be done with two priority values, or two "this is the gateway" flags. So, that doesn't help!

You can still split your network into a v4-only, and a v6-only network for that. I think we shouldn't go too far trying to handle edge-cases for an option that's not going to be used by the masses anyway.